### PR TITLE
Feature/custom transformers

### PR DESCRIPTION
**Pull request by In48semenov**
## Description
Example notebooks for custom transformer-based models.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Optimization

## How Has This Been Tested?
Before submitting a PR, please check yourself against the following list. It would save us quite a lot of time.
- Have you read the contribution guide?
- Have you updated the relevant docstrings? We're using Numpy format, please double-check yourself
- Does your change require any new tests?
- Have you updated the changelog file?

### CODE REVIEW

File: rectools/models/nn/transformer_base.py [hunk 1]
```diff
@@ -294,20 +301,84 @@ def _init_trainer(self) -> None:
...
+        Returns
+        -------
+        ItemNetConstructorBase
+            Network for item embeddings.
+        """
+        return self.item_net_constructor_type.from_dataset_schema(
+            dataset_schema, self.n_factors, self.dropout_rate, self.item_net_block_types
+        )
+
+    def init_pos_encoding_layer(self) -> PositionalEncodingBase:
```
Comment (blondered) [inline rectools/models/nn/transformer_base.py:340]:
```suggestion
    def _init_pos_encoding_layer(self) -> PositionalEncodingBase:
```

File: rectools/models/nn/transformer_base.py [hunk 2]
```diff
@@ -294,20 +301,84 @@ def _init_trainer(self) -> None:
...
+        Construct positional encoding layer.
+
+        Returns
+        -------
+        PositionalEncodingBase
+            Positional encoding layer.
+        """
+        return self.pos_encoding_type(self.use_pos_emb, self.session_max_len, self.n_factors)
+
+    def init_transformer_layers(self) -> TransformerLayersBase:
```
Comment (blondered) [inline rectools/models/nn/transformer_base.py:351]:
```suggestion
    def _init_transformer_layers(self) -> TransformerLayersBase:
```

File: rectools/models/nn/transformer_base.py [hunk 3]
```diff
@@ -294,20 +301,84 @@ def _init_trainer(self) -> None:
...
             n_blocks=self.n_blocks,
             n_factors=self.n_factors,
             n_heads=self.n_heads,
-            session_max_len=self.session_max_len,
             dropout_rate=self.dropout_rate,
-            use_pos_emb=self.use_pos_emb,
+        )
+
+    def _init_torch_model(self, dataset_schema: tp.Optional[DatasetSchema] = None) -> TransformerTorchBackbone:
+        if dataset_schema is None:
```
Comment (blondered) [inline rectools/models/nn/transformer_base.py:368]: This is not very clear.
Let's construct item net first. And then pass it to `_init_torch_model` method.

File: rectools/models/nn/sasrec.py [hunk 4]
```diff
@@ -137,15 +205,16 @@ def __init__(
...
         super().__init__()
         self.n_blocks = n_blocks
-        self.multi_head_attn = nn.ModuleList(
-            [torch.nn.MultiheadAttention(n_factors, n_heads, dropout_rate, batch_first=True) for _ in range(n_blocks)]
-        )  # important: original architecture had another version of MHA
-        self.q_layer_norm = nn.ModuleList([nn.LayerNorm(n_factors) for _ in range(n_blocks)])
-        self.ff_layer_norm = nn.ModuleList([nn.LayerNorm(n_factors) for _ in range(n_blocks)])
-        self.feed_forward = nn.ModuleList(
-            [PointWiseFeedForward(n_factors, n_factors, dropout_rate, torch.nn.ReLU()) for _ in range(n_blocks)]
+        self.transformer_layers = nn.ModuleList(
```
Comment (blondered) [inline rectools/models/nn/sasrec.py:208]:
```suggestion
        self.transformer_blocks = nn.ModuleList(
```
Otherwise we keep same naming for different things and it's not too clear.

File: rectools/models/nn/transformer_backbone.py [hunk 5]
```diff
@@ -63,64 +50,23 @@ class TransformerTorchBackbone(torch.nn.Module):
 
     def __init__(
         self,
-        n_blocks: int,
-        n_factors: int,
-        n_heads: int,
-        session_max_len: int,
+        n_transformer_heads: int,
```
Comment (blondered) [inline rectools/models/nn/transformer_backbone.py:53]: If you change naming, let'ы be more precise: `n_attn_heads`

Or just don't change it.
TransformerTorchBackbone doesn't include tying and targets. So we can't have any other heads except attention heads. So we can just keep it as `n_heads`

Multiple target heads will only be in lightning module.

File: rectools/models/nn/sasrec.py [hunk 6]
```diff
@@ -112,6 +112,74 @@ def _collate_fn_recommend(self, batch: List[Tuple[List[int], List[float]]]) -> D
         return {"x": torch.LongTensor(x)}
 
 
+class SASRecTransformerLayer(TransformerLayersBase):
```
Comment (blondered) [inline rectools/models/nn/sasrec.py:115]: About the notebook: 

1. "# For Bert-like models need add MASK to `_collate_fn` methods"
We really don't care about the mask at this point. NextToken is neither BERT-like (not MLM) or SASRec-like (shifted sequence).
It just predicts one next token. Use "MASK" (or correctly it would be "CLS") token or use just last item token representation for predictions - is really a very small detail.

2. `use_causal_attn=True` for NextItem doesn't change anything. SASRec has it by default. So it should be `use_causal_attn=False` to make a difference.

3. Just a fun fact: NextAction is extremely pop-biased. Makes sence. Great that we found it.

4. You can skip passing trainer funcs. These are all default params. And they are also all the same

Overall: I love the final result of transformers configuration and customization. Great job from all of us.

File: rectools/models/nn/transformer_base.py [hunk 7]
```diff
@@ -294,20 +301,84 @@ def _init_trainer(self) -> None:
         else:
             self._trainer = self.get_trainer_func()
 
-    def _init_torch_model(self) -> TransformerTorchBackbone:
-        return TransformerTorchBackbone(
+    def init_item_net_from_dataset(self, dataset: Dataset) -> ItemNetConstructorBase:
```
[the remark this answers is not in the document]
Comment (In48semenov) [reply rectools/models/nn/transformer_base.py]: fixed

File: rectools/models/nn/transformer_base.py (hunk 1 above)
Comment (In48semenov) [reply rectools/models/nn/transformer_base.py]: fixed

File: rectools/models/nn/transformer_base.py (hunk 2 above)
Comment (In48semenov) [reply rectools/models/nn/transformer_base.py]: fixed

File: rectools/models/nn/transformer_base.py (hunk 3 above)
Comment (In48semenov) [reply rectools/models/nn/transformer_base.py]: fixed

File: rectools/models/nn/sasrec.py (hunk 4 above)
Comment (In48semenov) [reply rectools/models/nn/sasrec.py]: fixed

File: rectools/models/nn/transformer_backbone.py (hunk 5 above)
Comment (In48semenov) [reply rectools/models/nn/transformer_backbone.py]: fixed

File: rectools/models/nn/sasrec.py (hunk 6 above)
Comment (In48semenov) [reply rectools/models/nn/sasrec.py:115]: Updated the tutorial.
If everything is ok, I will remove it and put it in the next PR

Comment (blondered) [reply rectools/models/nn/sasrec.py:115]: Thank you! Let's remove it and put in the next PR

File: rectools/models/nn/transformer_backbone.py [hunk 8]
```diff
@@ -167,9 +107,9 @@ def _merge_masks(
 
         merged_mask = attn_mask_expanded + key_padding_mask_expanded
         res = (
-            merged_mask.view(batch_size, 1, seq_len, seq_len)
+            merged_mask.reshape(batch_size, 1, seq_len, seq_len)
```
Comment (blondered) [inline rectools/models/nn/transformer_backbone.py:110]: Can we keep `view` here?

File: rectools/models/nn/sasrec.py [hunk 9]
```diff
@@ -177,19 +246,8 @@ def forward(
...
-                q, seqs, seqs, attn_mask=attn_mask, key_padding_mask=key_padding_mask, need_weights=False
-            )
-            seqs = q + mha_output
-            ff_input = self.ff_layer_norm[i](seqs)
-            seqs = self.feed_forward[i](ff_input)
-            seqs = self.dropout[i](seqs)
-            seqs += ff_input
-            seqs *= timeline_mask
-
+            seqs = self.transformer_blocks[i](seqs, timeline_mask, attn_mask, key_padding_mask)
```
Comment (blondered) [inline rectools/models/nn/sasrec.py:249]: We can stop passing `timeline_mask` to transformer_blocks (here and everywhere). Gonna be more clear.

```
for i in range(self.n_blocks):
  seqs *= timeline_mask
  seqs = self.transformer_blocks[i](seqs, attn_mask, key_padding_mask)
seqs *= timeline_mask
seqs = self.last_layernorm(seqs)
```

File: rectools/models/nn/sasrec.py (hunk 6 above)
Comment (In48semenov) [reply rectools/models/nn/sasrec.py]: Removed

File: rectools/models/nn/transformer_backbone.py (hunk 8 above)
Comment (In48semenov) [reply rectools/models/nn/transformer_backbone.py]: Fixed

File: rectools/models/nn/sasrec.py (hunk 9 above)
Comment (In48semenov) [reply rectools/models/nn/sasrec.py]: Fixed

**Review by blondered — APPROVED** [review]
